solver/src/main/java/org/chocosolver/solver/variables/view/integer/IntAffineView.java

```diff
@@ -0,0 +1,588 @@
+        } else {
+            _riterator.topDownInit();
+        }
+        return _riterator;
+    }
+
+    @Override
+    public boolean contains(int value) {
+        value -= b;
+        if (a > 1) {
+            if ((value % a) != 0) {
+                return false;
+            }
+            value /= a;
+        }
+        value = (!p) ? -value : value;
+        return var.contains(value);
+    }
+
+    @Override
+    public boolean isInstantiatedTo(int value) {
+        return var.isInstantiated() && contains(value);
+    }
+
+    @Override
+    public int getValue() throws IllegalStateException {
+        if (!isInstantiated()) {
+            throw new IllegalStateException("getValue() can be only called on instantiated variable. " +
+                    name + " is not instantiated");
+        }
+        return (p ? var.getValue() : -var.getValue()) * a + b;
+    }
+
+    @Override
+    public int getLB() {
+        return (!p ? -var.getUB() : var.getLB()) * a + b;
+    }
+
+    @Override
+    public int getUB() {
+        return (!p ? -var.getLB() : var.getUB()) * a + b;
+    }
+
+    @Override
+    public int nextValue(int v) {
+        if (v < getLB()) return getLB();
+        if (v > getUB()) return Integer.MAX_VALUE;
+        // y = p * a * x + b where p in {-1,1}, a and b in Z.
+        // => x = (y - b) / (p * a)
+        v -= b;
+        v = (v < 0 && a > 1) ? v - 1 : v;
+        v /= (p ? a : -a);
+
+        if (p) {
+            v = var.nextValue(v);
+        } else {
+            v = var.previousValue(v);
+        }
+        if (v == Integer.MIN_VALUE || v == Integer.MAX_VALUE) return Integer.MAX_VALUE;
+        return (p ? v : -v) * a + b;
+    }
+
+    @Override
+    public int nextValueOut(int v) {
+        v++;
+        if (v < getLB() || v > getUB()) return v;
+        // y = p * a * x + b where p in {-1,1}, a and b in Z.
+        // => x = (y - b) / (p * a)
+        double w = v - b;
+        w /= (p ? a : -a);
+
+        int k = (int) w;
+        if (w == k && var.contains(k)) {
+            if (a > 1) {
+                v++;
+            } else {
+                if (p) {
+                    k = var.nextValueOut(k);
+                } else {
+                    k = var.previousValueOut(k);
+                }
+                v = ((p ? k : -k) * a + b);
+            }
+        }
+        return v;
+    }
+
+    @Override
+    public int previousValue(int v) {
+        if (v > getUB()) return getUB();
+        if (v < getLB()) return Integer.MIN_VALUE;
+        // y = p * a * x + b where p in {-1,1}, a and b in Z.
+        // => x = (y - b) / (p * a)
+        v -= b;
+        v = (v > 0 && a > 1) ? v + 1 : v;
+        v /= (p ? a : -a);
+
+        if (p) {
+            v = var.previousValue(v);
+        } else {
+            v = var.nextValue(v);
+        }
+        if (v == Integer.MIN_VALUE || v == Integer.MAX_VALUE) return Integer.MIN_VALUE;
+        return (p ? v : -v) * a + b;
+    }
```

Comment:
This function is not correct. I tested it with the following code : 
`Model model = new Model();
 IntVar x = model.intVar(2, 9);
 IntVar y = model.intView(x, 5, 7);`

And calling `previousValue` has a "shifted" behaviour. For instance, `previousValue(51)` and `previousValue(52)` return 47 as they should, but `previousValue(48)`, `previousValue(49)` and `previousValue(50)` return 42 instead of 47 (and this "shift" is repeated when looping down the values).

This unwanted behaviour is due to the division done in line 485. Indeed, line 484 is not enough to compensate for it. Replacing line 484 by `v = (v > 0 && a > 1 && v % a != 0) ? v + a : v;` fixes it.


I think we should add a test checking this behaviour, assuring that from a value `v` such that `v % a == 0`, the `a` values above (from `v` to `v + (a - 1)`) all return `v + a` as `nextValue`, and the `a` values below (from `v - a + 1` to `v`) all return `v - a` as `previousValue`.